coreneuron/gpu/nrn_acc_manager.cpp

```diff
@@ -40,9 +41,227 @@ void nrn_ion_global_map_delete_from_device();
 void nrn_VecPlay_copyto_device(NrnThread* nt, void** d_vecplay);
 void nrn_VecPlay_delete_from_device(NrnThread* nt);
 
+int cnrn_target_get_num_devices() {
+#if defined(CORENEURON_ENABLE_GPU) && !defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && \
+    defined(_OPENACC)
+    // choose nvidia GPU by default
+    acc_device_t device_type = acc_device_nvidia;
+    // check how many gpu devices available per node
+    return acc_get_num_devices(device_type);
+#elif defined(CORENEURON_ENABLE_GPU) && defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && \
+    defined(_OPENMP)
+    return omp_get_num_devices();
+#else
+    throw std::runtime_error(
+        "cnrn_target_get_num_devices() not implemented without OpenACC/OpenMP and gpu build");
+#endif
+}
+
+void cnrn_target_set_default_device(int device_num) {
+#if defined(CORENEURON_ENABLE_GPU) && !defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && \
+    defined(_OPENACC)
+    acc_set_device_num(device_num, acc_device_nvidia);
+#elif defined(CORENEURON_ENABLE_GPU) && defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && \
+    defined(_OPENMP)
+    omp_set_default_device(device_num);
+    // It seems that with NVHPC 21.9 then only setting the default OpenMP device
+    // is not enough: there were errors on some nodes when not-the-0th GPU was
+    // used. These seemed to be related to the NMODL instance structs, which are
+    // allocated using cudaMallocManaged.
+    auto const cuda_code = cudaSetDevice(device_num);
+    assert(cuda_code == cudaSuccess);
```

Comment:
Would be nice to check if this is still needed with 21.11.

Reply:
It's still needed. I tried removing these two lines with the NVHPC 21.11 / CUDA 11.5.1 / GCC 11.2.0 / ... toolchain from the new deployment and saw several failures:
```
          6 - ring_TEST (Failed)
          7 - ring_binqueue_TEST (Failed)
          8 - ring_multisend_TEST (Failed)
          9 - ring_spike_buffer_TEST (Failed)
         10 - ring_permute1_TEST (Failed)
         11 - ring_permute2_TEST (Failed)
         12 - ring_gap_TEST (Failed)
         13 - ring_gap_binqueue_TEST (Failed)
         14 - ring_gap_multisend_TEST (Failed)
         15 - ring_gap_permute1_TEST (Failed)
         16 - ring_gap_permute2_TEST (Failed)
         17 - ring_permute2_cudaInterface_TEST (Failed)
         18 - ring_gap_permute2_cudaInterface_TEST (Failed)
         55 - external_nrntest (Failed)
         67 - external_ringtest::coreneuron_gpu_mpi_offline (Failed)
         94 - testcorenrn_gf::coreneuron_gpu_offline (Failed)
        110 - testcorenrn_vecplay::coreneuron_gpu_offline (Failed)
        130 - testcorenrn_watch::coreneuron_gpu_offline (Failed)
```
with these lines, I just see one failure with the same setup:
```
external_nrntest
```
presumably because this one is overly sensitive to compiler versions (https://github.com/neuronsimulator/nrntest/issues/22).